include/mfmg/hierarchy.hpp

```diff
@@ -0,0 +1,143 @@
+/*************************************************************************
+ * Copyright (c) 2017 by the mfmg authors                                *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * This file is part of the mfmg libary. mfmg is distributed under op BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#ifndef MFMG_HIERARCHY_HPP
+#define MFMG_HIERARCHY_HPP
+
+#include <memory>
+#include <vector>
+
+#include <boost/property_tree/ptree.hpp>
+
+namespace mfmg
+{
+
+template <class Operator>
+class Level
+{
+public:
+  std::shared_ptr<const Operator> get_operator() const { return _operator; }
+  std::shared_ptr<const Operator> get_restrictor() const { return _restrictor; }
+  std::shared_ptr<const Operator> get_prolongator() const
+  {
+    return _prolongator;
+  }
+  std::shared_ptr<const Operator> get_smoother() const { return _smoother; }
+
+  void set_operator(std::shared_ptr<const Operator> op) { _operator = op; }
+  void set_restrictor(std::shared_ptr<const Operator> r) { _restrictor = r; }
+  void set_prolongator(std::shared_ptr<const Operator> p) { _prolongator = p; }
+  void set_smoother(std::shared_ptr<const Operator> s) { _smoother = s; }
+
+private:
+  std::shared_ptr<const Operator> _operator, _prolongator, _restrictor,
+      _smoother;
+};
+
+template <class MeshEvaluator, class Operator, class Vector>
```

Comment:
Can we use `VectorType`? deal has already a class called `Vector` so it might lead to some confusion.